akka-actor/src/main/scala/akka/util/ImmutableIntMap.scala

```diff
@@ -3,139 +3,128 @@
  */
 package akka.util
 import java.util.Arrays
-
 import akka.annotation.InternalApi
-
 import scala.annotation.tailrec
 
 /**
  * INTERNAL API
  */
 @InternalApi private[akka] object ImmutableIntMap {
-  val empty: ImmutableIntMap =
-    new ImmutableIntMap(Array.emptyIntArray, Array.empty)
-
-  private final val MaxScanLength = 10
+  final val empty: ImmutableIntMap = new ImmutableIntMap(Array.emptyIntArray, 0)
 }
 
 /**
  * INTERNAL API
- * Specialized Map for primitive `Int` keys to avoid allocations (boxing).
- * Keys and values are backed by arrays and lookup is performed with binary
- * search. It's intended for rather small (<1000) maps.
+ * Specialized Map for primitive `Int` keys and values to avoid allocations (boxing).
+ * Keys and values are encoded consecutively in a single Int array and does copy-on-write with no
+ * structural sharing, it's intended for rather small maps (<1000 elements).
  */
-@InternalApi private[akka] final class ImmutableIntMap private (
-  private val keys: Array[Int], private val values: Array[Int]) {
-
-  final val size: Int = keys.length
+@InternalApi private[akka] final class ImmutableIntMap private (private final val kvs: Array[Int], final val size: Int) {
+
+  private[this] final def indexForKey(key: Int): Int = {
+    // Custom implementation of binary search since we encode key + value in consecutive indicies.
+    // We do the binary search on half the size of the array then project to the full size.
+    // >>> 1 for division by 2: https://research.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html
+    @tailrec def find(lo: Int, hi: Int): Int =
+      if (lo <= hi) {
+        val lohi = lo + hi // Since we search in half the array we don't need to div by 2 to find the real index of key
+        val idx = lohi & ~1 // We simply need to round down lo+hi by removing the lowest bit if set
+        val k = kvs(idx)
+        if (k < key) find((lohi >>> 1) + 1, hi)
+        else if (k > key) find(lo, (lohi >>> 1) - 1)
+        else idx
+      } else -((lo << 1) + 1) // Item should be placed in lo*2+1, negate it to indicate no match
+
+    find(0, size - 1)
+  }
 
   /**
    * Worst case `O(log n)`, allocation free.
+   * Will return Int.MinValue if not found, so beware of storing Int.MinValues
    */
-  def get(key: Int): Int = {
-    val i = Arrays.binarySearch(keys, key)
-    if (i >= 0) values(i)
-    else Int.MinValue // cant use null, cant use OptionVal, other option is to throw an exception...
+  final def get(key: Int): Int = {
+    // See algorithm description in `indexForKey`
```

Comment:
Too bad these binary searches can't be composed in like `Arrays.binarySearch` used to be. These two impls are nearly identical until their last two instructions. Would be cool to see reusable code instead of copypasta.

In any case, pardon my interruption, I'm just a drive-by twitter follower! ;)